jstests/core/hidden_index.js

```diff
@@ -0,0 +1,113 @@
+// during dml and ttl function works as expected.
+
+(function () {
+    'use strict';
+    load("jstests/libs/analyze_plan.js");  // For getPlanStages.
+    const testtab = db[jsTest.name()];
+
+    function usedIndex(query) {
+	printjson(testtab.getIndexes());
+	const explain = assert.commandWorked(testtab.find(query).explain());
+	printjson(explain);
+	const ixScans = getPlanStages(explain.queryPlanner.winningPlan, "IXSCAN");
+	return ixScans.length === 1;
+    }
+
+    function HiddenCheckTest(query, index_type, wildcard) {
+	let index_name;
+	if (wildcard)
+	    index_name = 'a.$**_' + index_type;
+	else
+	    index_name = 'a_' + index_type;
+
+	testtab.dropIndexes();
+	if (wildcard)
+	    assert.commandWorked(testtab.createIndex({"a.$**": index_type}));
+	else
+	    assert.commandWorked(testtab.createIndex({"a": index_type}));
+	
+	assert.eq(testtab.getIndexes()[1].hidden, undefined);
+	assert(usedIndex(query));
+	
+	assert.commandWorked(testtab.hideIndex(index_name));
+	assert(testtab.getIndexes()[1].hidden);
+	if (index_type === "text") {
+	    // errmsg: error processing query: ns=test.hidden_indexTree: TEXT : query=java,
+	    // language=english, caseSensitive=0, diacriticSensitive=0,
+	    // tag=NULL\nSort: {}\nProj: {}\n planner returned error ::
+	    // caused by :: need exactly one text index for $text query
+	    printjson(query);
+	    assert.throws(() => {testtab.find(query).explain();});
+	    return;
+	}
+	assert(!usedIndex(query));
+	
+	assert.commandWorked(testtab.unhideIndex(index_name));
+	assert.eq(testtab.getIndexes()[1].hidden, undefined);
+	assert(usedIndex(query));
+	
+	assert.commandWorked(testtab.dropIndex(index_name));
+
+	if (wildcard)
+	    assert.commandWorked(testtab.createIndex({"a.$**": index_type}, {hidden: true}));
+	else
+	    assert.commandWorked(testtab.createIndex({"a": index_type}, {hidden: true}));
+
+	assert(testtab.getIndexes()[1].hidden);
+	assert(!usedIndex(query));
+    }
+
+    
+    // normal index
+    HiddenCheckTest({a: 1}, 1);
+    // geo index
+    HiddenCheckTest({a: { $geoWithin :
+			  { $geometry :
+			    { type : "Polygon" ,
+                              coordinates : [ [
+                                  [ 0 , 0 ] ,
+                                  [ 3 , 6 ] ,
+                                  [ 6 , 1 ] ,
+                                  [ 0 , 0 ]
+                              ] ]
+			    } } } }, "2dsphere");
+
+    // fts index
+    HiddenCheckTest({ $text: { $search: "java" } }, "text");
+
+    // wildcard index
+    HiddenCheckTest({"a.f": 1}, 1, true);
+
+    // hidden index on capped collection
+    assert(testtab.drop());
+    assert.commandWorked(db.createCollection(jsTest.name(), {capped: true, size: 100}));
+    HiddenCheckTest({a: 1}, 1);
+    
+    // can't hide index on system index
+    const systemtab = db.getSiblingDB('admin').system.version;
+    assert.commandWorked(systemtab.createIndex({a: 1}));
+    assert.commandFailed(systemtab.hideIndex("a_1"));
```

Comment:
Can you substitute the assert.commandFailed() calls in this section with assert.commandFailedWithCode()? You would then pass the error code as a second argument. This is useful in that it will alert us if the reason a test fails has changed.